lib/SILOptimizer/Mandatory/DiagnoseInfiniteRecursion.cpp

```diff
@@ -0,0 +1,242 @@
+                WMI->getConformance(), WMI->getMember());
+            if (F == Target)
+              return true;
+
+            continue;
+          }
+        }
+      }
+      return false;
+    }
+
+    using ExitBlockSet = llvm::SmallPtrSetImpl<SILBasicBlock *>;
+    static bool hasInfinitelyRecursiveApply(SILFunction &Fn,
+                                            SILFunction *TargetFn,
+                                            const ExitBlockSet &ExitBlocks) {
+      enum RecursiveState {
+        FoundRecursion          = 0,
+        FoundPathOutOfFunction  = 1,
+      };
+
+      // Set the entry block into the state map.
+      llvm::DenseMap<SILBasicBlock *, RecursiveState> States;
+      States[Fn.getEntryBlock()] = FoundPathOutOfFunction;
+
+      // Make the processing stack and seed it with the entry block.
+      SmallVector<SILBasicBlock *, 16> Stack;
+      Stack.push_back(Fn.getEntryBlock());
+
+      while (!Stack.empty()) {
+        SILBasicBlock *CurBlock = Stack.pop_back_val();
+        RecursiveState &CurState = States[CurBlock];
+
+        if (CurState == FoundPathOutOfFunction) {
+          // Only change state if the block has a recursive call.
+          if (hasRecursiveCallInPath(*CurBlock, TargetFn)) {
+            CurState = FoundRecursion;
+          } else if (ExitBlocks.count(CurBlock)) {
+            // Found a path to the exit node without a recursive call.
+            return false;
+          }
+        }
+
+        // Loop over successor blocks and add them to the Stack if their state
+        // changes.
+        auto Iter = CurBlock->succ_begin();
+        for (auto EndIter = CurBlock->succ_end(); Iter != EndIter; ++Iter) {
+          if (SILBasicBlock *Succ = *Iter) {
+            if (States[Succ] < CurState) {
+              States[Succ] = CurState;
+              Stack.push_back(Succ);
+            }
+          }
+        }
+      }
+
+      // Bail out if all exit nodes are reachable, and only reachable, through
+      // a recursive call.
+      for (auto *EB : ExitBlocks) {
+        if (States[EB] == FoundPathOutOfFunction)
+          return false;
+      }
+      return true;
+    }
+
+    static SILFunction *computeLazyGetterClosure(PostOrderFunctionInfo *PO) {
+      for (auto *BB : PO->getReversePostOrder()) {
+        for (auto &I : *BB) {
+          auto *AI = dyn_cast<ApplyInst>(&I);
+          if (!AI) {
+            continue;
+          }
+
+          auto *Callee = AI->getCalleeFunction();
+          if (!Callee) {
+            continue;
+          }
+
+          // Ignore external declarations.
+          if (Callee->empty() || Callee->isExternalDeclaration()) {
+            continue;
+          }
+
+          // Ignore non-closure callees.
+          if (!Callee->getLocation().isASTNode<AbstractClosureExpr>()) {
+            continue;
+          }
+
+          return Callee;
+        }
+      }
+      return nullptr;
+    }
+
+    void run() override {
```

Comment:
Stylistically, I don't like nested method definitions more than a few lines long. (You ask me to review, you get my personal preferences too!)

Reply:
The bulk of this is the condition in and around `computeLazyGetterClosure` which can be factored out.  I had a question for @swiftix  about a better way to structure this that I'll ask you: The goal is to weed out as many non-lazy-var-initializer-looking closures as possible so I don't run off computing RPO info for them.  There's got to be a better condition than `Fn->isBare() == IsNotBare`.  Got any ideas?